docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,201 @@
+`onTimeoutPacket` callbacks, and verifies the timeout proof.
+
+The mechanism is similar to the transaction handling logic in `baseapp`. After
+authentication, the handler is executed, and the authentication state change
+must be committed regardless of the result of the handler execution.
+
+`handlePacketRecv` also requires acknowledgement writing which has to be done
+after the handler execution and also must be commited regardless of the result of
+the handler execution.
+
+## Decision
+
+The Cosmos SDK will define an `AnteHandler` for IBC packet receiving. The
+`AnteHandler` will iterate over the messages included in the transaction, type
+`switch` to check whether the message contains an incoming IBC packet, and if so
+verify the Merkle proof.
+
+```go
+// Pseudocode
+type ProofVerificationDecorator struct {
+  clientKeeper ClientKeeper
+  channelKeeper ChannelKeeper
+}
+
+func (pvr ProofVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  for _, msg := range tx.GetMsgs() {
+    var err error
+    switch msg := msg.(type) {
+    case client.MsgUpdateClient:
+      err = pvr.clientKeeper.UpdateClient(msg.ClientID, msg.Header)
+    case channel.MsgPacket:
+      err = pvr.channelKeeper.VerifyPacket(msg.Packet, msg.Proofs, msg.ProofHeight)
+    case chanel.MsgAcknowledgement:
+      err = pvr.channelKeeper.VerifyAcknowledgement(msg.Acknowledgement, msg.Proof, msg.ProofHeight)
+    case channel.MsgTimeoutPacket:
+      err = pvr.channelKeeper.VerifyTimeout(msg.Packet, msg.Proof, msg.ProofHeight, msg.NextSequenceRecv)
+    default:
+      continue
+    }
+
+    if err != nil {
+      return ctx, err.Result()
+    }
+  }
+  return next(ctx, tx, simulate)
+}
+```
+
+Where `MsgUpdateClient`, `MsgPacket`, `MsgAcknowledgement`, `MsgTimeoutPacket`
+are `sdk.Msg` types correspond to `handleUpdateClient`, `handleRecvPacket`,
+`handleAcknowledgementPacket`, `handleTimeoutPacket` of the routing module,
+respectively.
+
+The `ProofVerificationDecorator` will be inserted to the top level application.
+It should come right after the default sybil attack resistent layer from the
+current `auth.NewAnteHandler`:
+
+```go
+// add IBC ProofVerificationDecorator to the Chain of
+func NewAnteHandler(
+  ak keeper.AccountKeeper, supplyKeeper types.SupplyKeeper, ibcKeeper ibc.Keeper,
+  sigGasConsumer SignatureVerificationGasConsumer) sdk.AnteHandler {
+  return sdk.ChainAnteDecorators(
+    NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
+    ...
+    NewIncrementSequenceDecorator(ak),
+    ibcante.ProofVerificationDecorator(ibcKeeper.ClientKeeper, ibcKeeper.ChannelKeeper), // innermost AnteDecorator
+  )
+}
+```
+
+The Cosmos SDK will define the wrapper function `WriteAcknowledgement` under the
+ICS05 port keeper. The function will wrap packet handlers to automatically handle
+the acknowledgments.
+
+```go
+// Pseudocode
+type Ack struct {
```

Comment:
I don't think this is necessary. We can just use `sdk.Result` and, in case of an error, `sdk.ResultFromError()`